beacon-chain/sync/rpc_data_column_sidecars_by_range.go

```diff
@@ -0,0 +1,215 @@
+	"github.com/pkg/errors"
+
+	"github.com/sirupsen/logrus"
+)
+
+var notDataColumnsByRangeIdentifiersError = errors.New("not data columns by range identifiers")
+
+// dataColumnSidecarsByRangeRPCHandler looks up the request data columns from the database from a given start slot index
+func (s *Service) dataColumnSidecarsByRangeRPCHandler(ctx context.Context, msg interface{}, stream libp2pcore.Stream) error {
+	// We count a single request as a single rate limiting amount, regardless of the number of columns requested.
+	const rateLimitingAmount = 1
+
+	ctx, span := trace.StartSpan(ctx, "sync.DataColumnSidecarsByRangeHandler")
+	defer span.End()
+
+	// Check if the message type is the one expected.
+	request, ok := msg.(*pb.DataColumnSidecarsByRangeRequest)
+	if !ok {
+		return notDataColumnsByRangeIdentifiersError
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, respTimeout)
+	defer cancel()
+
+	SetRPCStreamDeadlines(stream)
+	beaconConfig := params.BeaconConfig()
+	maxRequestDataColumnSidecars := beaconConfig.MaxRequestDataColumnSidecars
+	remotePeer := stream.Conn().RemotePeer()
+
+	requestedColumns := request.Columns
+
+	// Format log fields.
+	var requestedColumnsLog interface{} = "all"
+	if uint64(len(requestedColumns)) != beaconConfig.NumberOfColumns {
+		requestedColumnsLog = requestedColumns
+	}
+
+	log.WithFields(logrus.Fields{
+		"remotePeer":       remotePeer,
+		"requestedColumns": requestedColumnsLog,
+		"startSlot":        request.StartSlot,
+		"count":            request.Count,
+	}).Debug("Serving data columns by range request")
+
+	// Validate the request regarding rate limiting.
+	if err := s.rateLimiter.validateRequest(stream, rateLimitingAmount); err != nil {
+		return errors.Wrap(err, "rate limiter validate request")
+	}
+
+	// Validate the request regarding its parameters.
+	rangeParameters, err := validateDataColumnsByRange(request, s.cfg.chain.CurrentSlot())
+	if err != nil {
+		s.writeErrorResponseToStream(responseCodeInvalidRequest, err.Error(), stream)
+		s.cfg.p2p.Peers().Scorers().BadResponsesScorer().Increment(remotePeer)
+		tracing.AnnotateError(span, err)
+		return errors.Wrap(err, "validate data columns by range")
+	}
+	if rangeParameters == nil {
+		return nil
+	}
+
+	// Ticker to stagger out large requests.
+	ticker := time.NewTicker(time.Second)
+
+	batcher, err := newBlockRangeBatcher(*rangeParameters, s.cfg.beaconDB, s.rateLimiter, s.cfg.chain.IsCanonical, ticker)
+	if err != nil {
+		s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+		tracing.AnnotateError(span, err)
+		return errors.Wrap(err, "new block range batcher")
+	}
+
+	// Derive the wanted columns for the request.
+	wantedColumns := make([]uint64, len(request.Columns))
+	copy(wantedColumns, request.Columns)
+
+	// Sort the wanted columns.
+	slices.Sort(wantedColumns)
+
+	var batch blockBatch
+	for batch, ok = batcher.next(ctx, stream); ok; batch, ok = batcher.next(ctx, stream) {
```

Comment:
what does this acheive chaining it like this? ( does it just keep going?)

Reply:
oh i see blobSidecarsByRangeRPCHandler is handled this way already

Reply:
> what does this acheive chaining it like this? ( does it just keep going?)
It keeps going until `batcher.next` returns `false`, which stops the `for` loop.